src/main/java/org/folio/circulation/infrastructure/storage/ServicePointRepository.java

```diff
@@ -197,6 +194,57 @@ public CompletableFuture<Result<MultipleRecords<Request>>> findServicePointsForR
           }));
   }
 
+  public CompletableFuture<Result<MultipleRecords<Request>>> findPrimaryServicePointsForRequests(
+    final MultipleRecords<Request> multipleRequests) {
+
+    log.debug("findServicePointsForRequests:: parameters multipleRequests: {}", () -> multipleRecordsAsString(multipleRequests));
+
+    final Collection<Request> requests = multipleRequests.getRecords();
+
+    final List<String> servicePointsToFetch = requests.stream()
+      .filter(Objects::nonNull)
+      .map(r -> r.getItem() != null ? r.getItem().getLocation().getPrimaryServicePointId() : null)
+      .filter(Objects::nonNull)
+      .map(Object::toString)
+      .distinct()
+      .toList();
+
+    if(servicePointsToFetch.isEmpty()) {
+      log.info("findPrimaryServicePointsForRequests:: No service points to query");
+      return completedFuture(succeeded(multipleRequests));
+    }
+
+    return createServicePointsFetcher().findByIds(servicePointsToFetch)
+      .thenApply(multipleServicePointsResult -> multipleServicePointsResult.next(
+        multipleServicePoints -> {
+          List<Request> newRequestList = new ArrayList<>();
+          Collection<ServicePoint> spCollection = multipleServicePoints.getRecords();
```

Comment:
I'd suggest putting some blank lines in between some of this code to make it easier to read.